beetmoverscript/tests/test_integration.py

```diff
@@ -0,0 +1,228 @@
+                },
+            ],
+        },
+    }
+
+
+def create_dummy_artifacts_for_task(task, root_path):
+    """Create artifacts to upload as specified in the given task."""
+    for artifact_dict in task["payload"]["artifactMap"]:
+        task_id = artifact_dict["taskId"]
+        for path in artifact_dict["paths"]:
+            full_path = root_path / task_id / path
+            full_path.parent.mkdir(parents=True, exist_ok=True)
+            create_dummy_artifact(full_path)
+
+
+def create_dummy_artifact(file_path):
+    with open(file_path, "wb") as f:
+        f.write(b"some data")
+
+
+@pytest.fixture
+def boto3_client_mock(monkeypatch):
+    client_mock = Mock(spec=beetmoverscript.script.boto3.client("s3"))
+    client_mock.generate_presigned_url.return_value = "presigned_url"
+
+    def fake_boto3_client_only_s3(*args, **kwargs):
+        assert args[0] == "s3", "A client other than 's3' was requested"
+        return client_mock
+
+    monkeypatch.setattr(beetmoverscript.script.boto3, "client", fake_boto3_client_only_s3)
+
+    return client_mock
+
+
+@pytest.fixture
+def aiohttp_session_mock(monkeypatch):
+    aiohttp_mock_data = {}
+
+    async def fake_text():
+        return "response_text"
+
+    response_mock = Mock(spec=beetmoverscript.script.aiohttp.ClientResponse)
+    response_mock.text = fake_text
+    response_mock.status = 200
+
+    @asynccontextmanager
+    async def fake_put(url, data=None, **kwargs):
+        aiohttp_mock_data.setdefault("put", []).append({"url": url, "data": data.read()})
+        yield response_mock
+
+    session_mock = Mock(spec=beetmoverscript.script.aiohttp.ClientSession)
+    session_mock.put = fake_put
+
+    @asynccontextmanager
+    async def fake_ClientSession(**kwargs):
+        yield session_mock
+
+    monkeypatch.setattr(beetmoverscript.script.aiohttp, "ClientSession", fake_ClientSession)
+    return aiohttp_mock_data
+
+
+@pytest.fixture
+def scriptworker_config(tmp_path, config, task):
+    work_dir = tmp_path / "work"
+    work_dir.mkdir()
+    config["work_dir"] = str(work_dir)
+
+    config_path = tmp_path / "config.json"
+    with open(config_path, "w") as config_file:
+        json.dump(config, config_file)
+
+    with open(work_dir / "task.json", "w") as task_file:
+        json.dump(task, task_file)
+
+    create_dummy_artifacts_for_task(task, work_dir / "cot")
+
+    return config_path
+
+
+def test_main_android_components_nightly(scriptworker_config, boto3_client_mock, aiohttp_session_mock):
+    main(config_path=scriptworker_config)
+
+    # Number of artifacts put matches expected
+    assert len(aiohttp_session_mock["put"]) == 12
+
+    # url to where artifacts are put is the presigned url
+    assert all(put_call["url"] == "presigned_url" for put_call in aiohttp_session_mock["put"])
+    # The data put matches the data in the files
+    assert all(put_call["data"] == b"some data" for put_call in aiohttp_session_mock["put"])
```

Comment:
Nit: Let's add another assertion about the 12 paths. They should match the paths requested in the task payload. This way, if we end up breaking these paths (beetmover used to generate on the fly), then this test will let us know.

Reply:
:+1: I already had in mind that adding some assertion about the paths would be a good idea. However, should we check source paths, destination paths, or even both?